utils/with-json-input.ts

```diff
@@ -0,0 +1,46 @@
+import * as t from "io-ts";
+import * as RA from "fp-ts/lib/ReadonlyArray";
+import * as E from "fp-ts/lib/Either";
+import { Context } from "@azure/functions";
+import { Json, JsonFromString } from "io-ts-types";
+import { pipe } from "fp-ts/lib/function";
+import { readableReport } from "@pagopa/ts-commons/lib/reporters";
+
+/**
+ * Wrap a function handler so that every input is a valid JSON object
+ * Useful to normalize input coming from queueTrigger, which could be bot a parsed object or a stringified object
+ *
+ * @param handler the handler to be executed
+ * @returns
+ */
+export const withJsonInput = <T = unknown>(
+  handler: (
+    context: Context,
+    ...parsedInputs: ReadonlyArray<Json>
+  ) => Promise<T>
+) => (context: Context, ...inputs: ReadonlyArray<unknown>): Promise<T> =>
+  pipe(
+    inputs,
+    RA.map(input =>
+      pipe(
+        input,
+        t.string.decode,
+        E.chain(JsonFromString.decode),
+        E.fold(_ => Json.decode(input), E.of)
+      )
+    ),
+    RA.sequence(E.Applicative),
+    E.getOrElseW(err => {
+      context.log.error(
+        `${
+          context.executionContext.functionName
+        }|invalid incoming queue item|${readableReport(err)}`
+      );
+      throw new Error(
+        `Cannot parse incoming queue item into JSON object: ${readableReport(
```

Comment:
won't throwing from a queue handler cause a retry?

Reply:
Good point. Yes it does, and of course every retry would result in the very same error. That would result in wasted executions, I see. Nevertheless, let's also say items are injected into queues from well-known, strongly typed producers; thus we may have compensated the issue by preventing it. 

Anyway, what a solution would be? To return the function _successfully_ but with a failure code? I would be fine with that except we would not see such failures into our common monitoring dashboards, we would need specific query for such events. 

As for now, my opinion is we may accept resource-wasting more than observability loss. What's you opinion?

Reply:
>  To return the function successfully but with a failure code?

Yep, that's what we do in other parts (ie. to store the return value of durable function).

>  I would be fine with that except we would not see such failures into our common monitoring dashboards

You can catch any exception for permanent errors (ie. make methods return either of `TransientError` or `PermanentError`) and avoid throwing (just throw in case of `TransientError`). You can manually call ApplicationInsight `trackException()` method to store it in AI dashboard (otherwise you'll get a lot of duplicated error logs during the retry).

Take a look at this old code:

https://github.com/pagopa/io-functions/blob/081e9caba58551eff930b19d3833f622b5708af3/lib/utils/azure_queues.ts#L184